wdl-doc/src/docs_tree.rs

```diff
@@ -0,0 +1,336 @@
+//! Implementations for a [`DocsTree`] which represents the DOCS directory.
+
+use std::collections::BTreeMap;
+use std::fs::canonicalize;
+use std::path::Path;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+use maud::Markup;
+use maud::html;
+use pathdiff::diff_paths;
+
+use crate::Document;
+use crate::full_page;
+use crate::r#struct::Struct;
+use crate::task::Task;
+use crate::workflow::Workflow;
+
+/// The type of a page.
+#[derive(Debug)]
+pub enum PageType {
+    /// An index page.
+    Index(Document),
+    /// A struct page.
+    Struct(Struct),
+    /// A task page.
+    Task(Task),
+    /// A workflow page.
+    Workflow(Workflow),
+}
+
+/// An HTML page in the DOCS directory.
+#[derive(Debug)]
+pub struct HTMLPage {
+    /// The display name of the page.
+    name: String,
+    /// The type of the page.
+    page_type: PageType,
+}
+
+impl HTMLPage {
+    /// Create a new HTML page.
+    pub fn new(name: String, page_type: PageType) -> Self {
+        Self { name, page_type }
+    }
+
+    /// Get the name of the page.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Get the type of the page.
+    pub fn page_type(&self) -> &PageType {
+        &self.page_type
+    }
+}
+
+/// A node in the DOCS directory tree.
+#[derive(Debug)]
+struct Node {
+    /// The name of the node.
+    name: String,
+    /// The absolute path to the node.
+    path: PathBuf,
+    /// The page associated with the node.
+    page: Option<Rc<HTMLPage>>,
+    /// The children of the node.
+    children: BTreeMap<String, Node>,
+}
+
+impl Node {
+    /// Create a new node.
+    pub fn new<P: Into<PathBuf>>(name: String, path: P) -> Self {
+        Self {
+            name,
+            path: path.into(),
+            page: None,
+            children: BTreeMap::new(),
+        }
+    }
+
+    /// Get the name of the node.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Get the path of the node.
+    pub fn path(&self) -> &PathBuf {
+        &self.path
+    }
+
+    /// Get the page associated with the node.
+    pub fn page(&self) -> Option<Rc<HTMLPage>> {
+        self.page.clone()
+    }
+
+    /// Iterate over the children of the node in a Depth First Traversal order.
+    pub fn depth_first_traversal(&self) -> Vec<&Node> {
```

Comment:
I think this function could be implemented with a local function that takes the `Vec` by mutable reference so that it doesn't have to allocate and deallocate a `Vec` at every node.

For example:

```rust
pub fn depth_first_traversal(&self) -> Vec<&Node> {
    fn recurse_depth_first(node: &Node, nodes: &mut Vec<&Node>) {
        nodes.push(self);

        for child in self.children.values() {
            recurse_depth_first(child, nodes);
        }
    }

    let mut nodes = Vec::new();
    recurse_depth_first(self, &mut nodes);

    nodes
}
```

Reply:
Alternatively, depending on how this method is used, we could not return a `Vec` and instead accept a `FnMut` callback that would "visit" each node in DFS order.

It certainly looks like the places we use this method could be translated to:

```rust
root.depth_first_traversal(|node| {
    // do something with node
});
```

Reply:
Although, looking at what `maud` supports, we probably can't use this within a `html!` macro, so returning a `Vec` is fine.